### PR TITLE
feat(gemma4): serve requests through the engine contract

### LINKED ISSUE #889: gemma4: serve 12B requests through the engine contract
## Description

Everything Gemma 4 needs to answer a request has landed piece by piece: the text tower loads, the tokenizer and chat template are gated against a reference, both attention families execute, the full forward matches a Hugging Face fixture, the KV pools serve both families, and the sliding window releases its front as the frontier moves. Nothing connects them to a request. `start_engine` bails with "not implemented yet (registration only)", so the model line probes a checkpoint and then refuses to launch it.

## Proposed Plan

- One owned thread serving requests serially through the KV-backed forward, with admission per step and a request whose prompt plus requested output exceeds the serving ceiling refused at the engine boundary rather than truncated.
- Next-token selection and per-token logprobs through the shared sampler on device logits.
- EOS and length finish reasons, with the stop token retiring the request instead of being emitted: the vLLM bridge appends its own sentinel for a terminal Stop and drops the last id, so an engine that emits EOS costs the client its final visible token.
- The ids the checkpoint lists under `suppress_tokens` dropped from contention before sampling and before the logprob normalization, read from the checkpoint rather than assumed for the family.
- Cancellation as the sink refusing an event: the request retires mid-stream and its KV returns by drop.
- The model line stops being registration only: it validates `--cuda-graph` before the weight load and publishes the context this engine can serve, so the frontend advertises that instead of the checkpoint's `max_position_embeddings`.

## Acceptance Criteria

- The direct engine stream matches the certified generate fixture token for token. All three fixture prompts return byte-identical text through `/v1/completions`, and both completions and chat work in streaming and non-streaming modes.
- A chat turn ends on the model's stop token. The engine emits no stop token, and its `completion_tokens` equals the number of emitted `Token` events.
- An over-length prompt is refused with the ceiling named, and the advertised context length is the one the engine can actually serve.
- A dropped stream retires the request without taking the engine down; the next request is served normally.

### PR DESCRIPTION
**Pull request by FeathBow**
## Description

Closes #889.

One owned thread serves Gemma 4 requests serially through the KV-backed forward. Admission runs per step, and a request whose prompt plus requested output exceeds the serving ceiling is refused at this boundary; next-token selection and per-token logprobs go through `pegainfer-sample` on device logits; EOS and length finish reasons are real from the first request. Cancellation is the sink refusing an event: the request retires mid-stream, its KV returns by drop, and the engine serves the next one.

The stop token retires a request without being emitted, which is the protocol the frontend documents: it appends its own sentinel for the terminal `Stop` and unconditionally drops the last id, so an engine that emits EOS costs the client its final visible token. The ids the checkpoint lists under `suppress_tokens` are dropped out of contention before anything reads the logits row, so sampling and the logprob normalization see the same distribution and a placeholder id can neither reach the client nor return as the next step's input. Both sets are read from `generation_config.json` rather than assumed for the family: 12B declares the end-of-image and end-of-audio ids, 26B and 31B declare none.

At the Gemma engine boundary, everything this stage does not implement is refused loudly rather than downgraded. At load: CUDA graph capture, multi-device ordinals, parallel topologies. Per request: a resolved KV prefix, `kv_transfer_params`, a nonzero `data_parallel_rank`, a LoRA adapter, prompt echo.

The model line stops being registration only. `--cuda-graph` defaults to true for every line, so an explicit `--cuda-graph=true` is refused in `ModelLine::validate` before the weight load, and everything else launches eagerly. The handle publishes an 8192-token serving ceiling, which is what lets the frontend reject an over-length prompt with its own message; without it the frontend advertises the checkpoint's 262144 `max_position_embeddings` and forwards a request the engine can only fail mid-stream.

That ceiling is what bounds the rope tables and the pool budget: each KV pool is sized for one full-context request plus its padding page, since a prompt is prefilled in one step and both families hold every position of it before the window releases anything. The 262k the checkpoint declares needs a table and KV budget design of its own.

### Test Env

Single GPU (sm_89, x86_64), CUDA 12.9, pinned 12B checkpoint.

### Verification

- fmt; clippy `-D warnings` over all gemma4 targets with and without the feature; the CI package set with `--locked`; the `pegainfer-kernels` hd256 smoke; gemma4 lib tests; `pegainfer-server` builds with `--features gemma4`.
- Checkpoint gates (`--ignored --test-threads=1`): 8 passed, including all three certified generate fixture prompts matching token for token.
- Live server: `/v1/models` reports `max_model_len` 8192. All three certified fixture prompts, sent as `/v1/completions` text at `temperature 0`, return text byte-identical to the fixture continuations (3/3), with prompt token counts matching the fixture (18/13/16).
- Live server, routes: completions and chat each return identical text streaming and non-streaming (`'111.1.'` over 8 chunks, and `Paris` respectively); the chat turn ends with `finish_reason=stop` and no stop token in the text; per-token logprobs come back with the picked token's logprob equal to its top entry; dropping a stream mid-decode aborts the request and the next one is served normally.
- Live server, length limits: a 9001-token prompt is refused in 0.1 s with HTTP 400 and `this model's maximum context length is 8192 tokens, but the prompt contains 9001 input tokens`. A one-token prompt asking for 9000 output tokens is not refused: the frontend clamps it to the remaining budget and the request completes with `finish_reason=length` at 8191 completion tokens.
- Live server, CLI: `--cuda-graph=true` exits with `Gemma 4 serves eagerly; --cuda-graph=true is not supported` and no weight-load line in its output; `--cuda-graph=false` serves.

## Type of Change

- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: pegainfer-gemma4/src/engine.rs [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+
+        let reject = |message: String| {
+            let _ = sink.send(TokenEvent::Rejected {
+                message,
+                prompt_tokens,
+                completion_tokens: 0,
+            });
+        };
+        if prompt_tokens == 0 {
+            return reject("empty prompt".into());
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply pegainfer-gemma4/src/engine.rs:251]: This is already enforced by GemmaServe::step(). It calls forward::validate_tokens() before planning, H2D, or embedding, and that shared validation rejects all six modality placeholder IDs.

**Comment by FeathBow** [comment]
@codex review please

File: pegainfer-gemma4/src/engine.rs [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+    let eos = token_ids(
+        json.get("eos_token_id")
+            .with_context(|| format!("{path} missing eos_token_id"))?,
+    )
+    .with_context(|| format!("{path} eos_token_id"))?;
+    anyhow::ensure!(!eos.is_empty(), "{path} declares an empty eos_token_id");
+    let suppress = match json.get("suppress_tokens") {
+        Some(value) => token_ids(value).with_context(|| format!("{path} suppress_tokens"))?,
+        None => Vec::new(),
+    };
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply pegainfer-gemma4/src/engine.rs]: Good catch. Prompt validation protects step inputs, but a generated placeholder is emitted before that validation runs again.

File: pegainfer-gemma4/src/model_line.rs [hunk 3]
```diff
@@ -45,9 +47,30 @@ impl ModelLine for Gemma4Line {
...
+    }
+
     fn launch(&self, ctx: &LaunchContext<'_>) -> anyhow::Result<LaunchedEngine> {
-        crate::start_engine(ctx.model_path, EngineLoadOptions::default())
-            .map(LaunchedEngine::Handle)
+        crate::start_engine(
+            ctx.model_path,
+            &EngineLoadOptions {
+                enable_cuda_graph: false,
+                ..EngineLoadOptions::default()
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply pegainfer-gemma4/src/model_line.rs:71]: Makes sense.

**Comment by FeathBow** [comment]
@codex review again please
